### PR TITLE
Fix integration tests caused by recent helm change

### PR DESCRIPTION
**Pull request by hklai**
Also adding prow jobs to run integration tests, and fixing codecov to
exclude unnecessary packages.

### CODE REVIEW

**Comment by nmittler** [comment]
@hklai `ci/circleci: test-integration-kubernetes ` is failing now?

**Comment by hklai** [comment]
@nmittler it actually failed even in yours (https://github.com/istio/istio/pull/9612)

**Comment by nmittler** [comment]
@hklai oh I thought it was passing before .. did someone change something on master that re-broke it?

**Comment by hklai** [comment]
I remember seeing that passed in your PR once, but not anymore.

I can't say for sure if any other recent change had broken it again. We
unfortunately don't have postsubmit testgrid set up for these tests.

File: tests/integration2/tests.mk [hunk 1]
```diff
@@ -29,39 +29,32 @@ ifneq ($(KUBECONFIG),)
...
+JUNIT_UNIT_TEST_XML ?= $(ISTIO_OUT)/junit_unit-tests.xml
+JUNIT_REPORT = $(shell which go-junit-report 2> /dev/null || echo "${ISTIO_BIN}/go-junit-report")
 
-# All integration tests targeting Kubernetes environment.
-.PHONY: test.integration.kube
-test.integration.kube: $(_INTEGRATION_TESTS_KUBE)
+# TODO: Exclude examples and qualification since they are very flaky.
+TEST_PACKAGES = $(shell go list ./tests/integration2/... | grep -v /qualification | grep -v /examples)
 
-# Generate integration test targets for local environment.
```
Comment (ozevren) [inline tests/integration2/tests.mk]: Please don't remove these targets. These create test.integration.component and test.integration.kube.component targets, which are very useful for running one-off integration tests.

File: tests/integration2/tests.mk [hunk 2]
```diff
@@ -29,39 +29,32 @@ ifneq ($(KUBECONFIG),)
...
-$(_INTEGRATION_TESTS_LOCAL): test.integration.%:
-	$(GO) test -p 1 ${T} ./tests/integration2/$*/... --istio.test.env local
+# All integration tests targeting local environment.
+test.integration: | $(JUNIT_REPORT)
+	mkdir -p $(dir $(JUNIT_UNIT_TEST_XML))
+	set -o pipefail; \
+	$(GO) test -p 1 ${T} ${TEST_PACKAGES} --istio.test.env local \
+	2>&1 | tee >($(JUNIT_REPORT) > $(JUNIT_UNIT_TEST_XML))
 
-# Generate integration test targets for kubernetes environment.
```
Comment (ozevren) [inline tests/integration2/tests.mk]: This one as well.

File: tests/integration2/tests.mk [hunk 3]
```diff
@@ -29,39 +29,32 @@ ifneq ($(KUBECONFIG),)
...
 
-# All integration tests targeting local environment.
-.PHONY: test.integration
-test.integration: $(_INTEGRATION_TESTS_LOCAL)
+JUNIT_UNIT_TEST_XML ?= $(ISTIO_OUT)/junit_unit-tests.xml
+JUNIT_REPORT = $(shell which go-junit-report 2> /dev/null || echo "${ISTIO_BIN}/go-junit-report")
 
-# All integration tests targeting Kubernetes environment.
-.PHONY: test.integration.kube
-test.integration.kube: $(_INTEGRATION_TESTS_KUBE)
```
Comment (ozevren) [inline tests/integration2/tests.mk]: This and test.integration are the ones that should be changed to support single output for junit integration.

File: pkg/test/deployment/helm.go [hunk 4]
```diff
@@ -108,39 +108,38 @@ func HelmTemplate(deploymentName, namespace, chartDir, workDir, valuesFile strin
...
+	if _, err := exec(fmt.Sprintf("helm --home %s package -u %s -d %s", helmRepoDir, chartDir, chartBuildDir)); err != nil {
+		return "", err
 	}
-	return nil
+	return exec(fmt.Sprintf("helm --home %s template %s --name %s --namespace %s%s%s",
+		helmRepoDir, chartDir, deploymentName, namespace, valuesFileString, valuesString))
 }
 
-func exec(cmd string) error {
+func exec(cmd string) (string, error) {
```
Comment (ozevren) [inline pkg/test/deployment/helm.go]: We have a shell.Execute in another package. Can we use it instead?

File: pkg/test/deployment/helm.go [hunk 5]
```diff
@@ -108,39 +108,38 @@ func HelmTemplate(deploymentName, namespace, chartDir, workDir, valuesFile strin
...
+		return "", err
 	}
 
-	// Create a dir for the packaged charts.
-	chartBuildDir, err := ioutil.TempDir(workDir, "helm-charts-")
-	if err != nil {
-		return err
+	// Adding cni dependency as a workaround for now.
+	if _, err := exec(fmt.Sprintf("helm --home %s repo add istio.io %s",
+		helmRepoDir, "https://raw.githubusercontent.com/istio/istio.io/master/static/charts")); err != nil {
```
Comment (ozevren) [inline pkg/test/deployment/helm.go]: Can't we consume this from the local enlistment? Wouldn't this break testing things in the local repo?

File: pkg/test/deployment/helm.go [hunk 6]
```diff
@@ -108,39 +108,38 @@ func HelmTemplate(deploymentName, namespace, chartDir, workDir, valuesFile strin
 		valuesFileString = fmt.Sprintf(" --values %s", valuesFile)
 	}
 
-	err := initCharts(chartDir, workDir)
-	if err != nil {
+	helmRepoDir := workDir + "/helmrepo"
```
Comment (ozevren) [inline pkg/test/deployment/helm.go]: Please use/add paths in pkg/test/env

File: pkg/test/deployment/helm.go (hunk 5 above)
Comment (hklai) [reply pkg/test/deployment/helm.go]: There is a new hard-coded dependency on this from https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/requirements.yaml#L60

Without this the following helm commands will fail. Istio package is built below from local workspace.

File: pkg/test/deployment/helm.go (hunk 6 above)
Comment (hklai) [reply pkg/test/deployment/helm.go]: workDir is actually a temp dir and helmRepo is created on top of it. This looks different from the other paths in istio/pkg/test/env/istio.go.

**Comment by hklai** [comment]
@ozevren Thanks. Replied and updated tests.mk.

File: pkg/test/deployment/helm.go (hunk 6 above)
Comment (nmittler) [reply pkg/test/deployment/helm.go]: nit: use `filepath.Join()` here and below.

File: prow/istio-integ-local-tests.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+WD=$(dirname "$0")
+WD=$(cd "$WD"; pwd)
+ROOT=$(dirname "$WD")
+
+# No unset vars, print commands as they're executed, and exit on any non-zero
+# return code
+set -u
```
Comment (nmittler) [inline prow/istio-integ-local-tests.sh]: nit: add the same comments that appear in the other script.

**Pull request merged**